jit pickling rref

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32959 jit pickling rref**

in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

zhaojuanmao committed:
jit pickling rref

in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit e08cae3d:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=e08cae3d0894eda05f90c719ddd01af6d5180173).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4407466) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94828830) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94828830))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
do so (now or later) by using -b with the checkout command again. Example: 
 
  git checkout -b <new-branch-name> 
 
HEAD is now at e08cae3d08 jit pickling rref 
+ git reset --hard e08cae3d0894eda05f90c719ddd01af6d5180173 
HEAD is now at e08cae3d08 jit pickling rref 
+ git merge --allow-unrelated-histories --no-edit --no-ff 5ca7bf453df9b5854062dc581ffd398f6b1b2b30 
Auto-merging torch/testing/_internal/distributed/rpc/rpc_test.py 
CONFLICT (content): Merge conflict in torch/testing/_internal/distributed/rpc/rpc_test.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

pritamdamania87 commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -145,6 +148,24 @@ void Pickler::pushDevice(const IValue& ivalue) {
+  auto rref = c10::static_intrusive_pointer_cast<distributed::rpc::RRef>(rrefInterface);
+  pushGlobal("torch", "rref");
+  auto& ctx = distributed::rpc::RRefContext::getInstance();
+  auto rrefForkData = ctx.prepareChildFork(rref);
```
I know this is what PyRRef does, but as the comment there mentions this approach is error prone. If we jit serialize a model containing an RRef we would end up creating a child fork incorrectly here.

Can we ensure we create a child fork only when we're sending RPCs?

@mrshenli Do we have gh issue tracking that we need to fix this for PyRRef?

mrshenli replied:
Just created one #32966

zhaojuanmao replied:
agree this is an issue we need to tackle. but I think it is better tackled with python pickling together, as in general JIT pickling serialization format should be same as Python pickling serialization format

pritamdamania87 replied:
This is a pretty serious bug during pickling and unpickling. Can we fix this the right way in this PR?

mrshenli replied:
One option would be having a thread_local flag/context. We use this context/flag to wrap the code below (and for remote as well, basically every rpc related entry point). Then the pickler can check the thread_local context/flag. If not set, throw an error.

https://github.com/pytorch/pytorch/blob/d613bd0522d51602ccf8cb29d00eb26cf3af923d/torch/distributed/rpc/api.py#L457-L465

pritamdamania87 replied:
Yanli and I spoke offline and thought its better to deal with this in a follow up PR. Although, lets fix this issue soon as a follow up PR for both jit serialization and PyRRef serialization.

The thread local idea sounds good since it can be reused across python pickle and JIT pickle.

zhaojuanmao replied:
sounds good to me as well

pritamdamania87 commented on torch/csrc/jit/unpickler.cpp:
```diff
@@ -503,6 +504,33 @@ void Unpickler::readGlobal(
       return;
+  } else if (module_name == "torch" && class_name == "rref") {
+      globals_.emplace_back([this] {
+        auto args = stack_.back().toTuple()->elements();
```
nit: Should we add an assert for the number of args we expect?

pritamdamania87 commented on torch/csrc/jit/unpickler.cpp:
```diff
@@ -503,6 +504,33 @@ void Unpickler::readGlobal(
+        stack_.emplace_back(std::move(c10::static_intrusive_pointer_cast<c10::RRefInterface>(rref)));
+      });
+      stack_.emplace_back(int64_t(globals_.size() - 1));
+      return;
```
Similar to the pickle method, this is error prone when we jit unpickle a module that contains an RRef. In general pickling/unpickling should not have any RRef fork logic, only the RPC Framework should have this logic.

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -177,7 +177,10 @@ c10::intrusive_ptr<RRef> RRefContext::getOrCreateRRef(
+    //  Comment it to unblock test_remote_script_module,
+    //  TODO, Will uncomment it when we are able to pass class resolver to JIT
+    //  unplicker and parse correct RRef type during JIT unpickling
+    //  TORCH_INTERNAL_ASSERT(ownerRRef->type() == type);
```
I think it might be fine to patch this PR for the performance effort to unblock them. But I don't think we should land a PR with a change like this where we remove an important assert.

zhaojuanmao replied:
agree, as commented here, this is mainly to unblock test, not for landing diff

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1722,34 +1722,36 @@ def rref_to_here(rref_var):
+        # ref as arg is passed to pybind boundary, and the ref is not garbage
+        # collected by python when calling shutdown()
+        import torch.distributed.rpc.api as api
+        api._ignore_rref_leak = True
```
@mrshenli How do other RRef tests guarantee that RRefs are garbage collected before shutdown() is called?

mrshenli replied:
Other tests rely on Python to deref local RRef objects when exiting the test function context, and then uses agent.join() to wait for all messages settle.

mrshenli commented:
@zhaojuanmao could you please confirm that the leak you encountered in this PR did not occur in #32674?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1722,34 +1722,36 @@ def rref_to_here(rref_var):
+
+        n = self.rank + 1
+        dst_rank = n % self.world_size
+        remote_ref = rpc.remote(
```
IIUC, this will return a PyRRef that contains an intrusive_ptr of RRef. So, when you pass this as an arg to `rpc.rpc_sync`, how does `createStackForSchema` work? Does it know it should skip the pickler of PyRRef?

mrshenli replied:
How does pickler serialize `PyObjectHolder`?

zhaojuanmao replied:
@mrshenli I checked the code, it worked like this:
when we define torchscript function run_ref_script_module(), its input type is annotated, so when we call createStackForSchema(), JIT knows the input type is RRef type, then when it reaches toIvalue of pybind_utils.h, it will treat the argument as RRef type and call obj.cast<torch::distributed::rpc::PyRRef>().toIValue() => this will return c10::static_intrusive_pointer_cast<c10::RRefInterface>(rref_), which is the IValue we want

zhaojuanmao commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -1,5 +1,6 @@
 #include <ATen/ATen.h>
 #include <ATen/core/Dict.h>
+#include <torch/csrc/distributed/rpc/rref_context.h>
```
So far the build and test do not complain, but if there is another approach to avoid dependency on rref_context, that is even better, will follow up with you offline

zhaojuanmao commented:
> @zhaojuanmao could you please confirm that the leak you encountered in this PR did not occur in #32674?

Yes, the leak did not happen in #32674

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

driazati commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -145,6 +154,28 @@ void Pickler::pushDevice(const IValue& ivalue) {
 }
 
+#ifdef USE_DISTRIBUTED
+void Pickler::pushRRef(const IValue& ivalue) {
```
I'm not very familiar with the semantics of RRef are but the changes to the pickler here look fine. Can you elaborate on what it means to serialize an RRef? Is it possible to load a model much later and have a dead reference or something?

zhaojuanmao replied:
thanks for reviewing @driazati. We are implementing how to serialize RRef during RPC call only right now, we have not implemented how to save a model with RRef. 

When we make RPC call that takes RRef as args, we need to pickle RRef by passing a few IDs and type of the RRef over the wire, and then recreate the RRef from the received IDs and type during unpickling.

I will add more description in the summary

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

suo commented:
Checking in here: what is the status of this? Is it blocked on waiting for review? If so, who is the best reviewer? Is it me? 😛

zhaojuanmao commented:
@suo thanks for checking. Yes, this diff is blocked by reviewing from both JIT team and our team. 

That would be great if you can help review whether you are OK with JIT unpickler constructor change, class_resolver does not work for our case, I added a type_resolver to JIT unpickler.  

@driazati commented he is fine with where I put rref pickling logic to.

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

zhaojuanmao commented:
updated summary in git commit message (did not show in the PR summary somehow) and updated some comments

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -673,6 +673,15 @@ std::unordered_map<std::string, std::string> ProcessGroupAgent::getMetrics() {
 }
 
+TypeResolver ProcessGroupAgent::getTypeResolver() {
+  auto typeResolver = [&](const std::string& typeStr) {
```
any reason for creating a new lambda every time calling this? This will be on the critical path as we need this for every message. Is it possible to have a static var here, and share it through `shared_ptr` or reference?

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -40,7 +29,9 @@ RRefForkData fromPyTuple(const py::tuple& pyTuple) {
       pyTuple.size() == RFD_TUPLE_SIZE,
-      "Pickled RRefForkData must contain 6 numbers.");
+      "Pickled RRefForkData must contain ",
+      RFD_TUPLE_SIZE,
```
Thanks for fixing this!

mrshenli commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -11,6 +11,11 @@ namespace torch {
 
+// Input is qualified name string or type str, output is JIT typePtr
+// Same as jit::TypeResolver, did not import jit::TypeResolver to here
+// because it could instroduce cyclic dependencies.
```
instroduce => introduce

mrshenli reviewed (commented):
For the leak, I set `api._ignore_rref_leak = False`, and run `test_remote_script_module` 60 times, but cannot reproduce the leak. Do you still see the leak in your dev env?

pritamdamania87 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -673,6 +673,15 @@ std::unordered_map<std::string, std::string> ProcessGroupAgent::getMetrics() {
   return metrics;
 }
 
+TypeResolver ProcessGroupAgent::getTypeResolver() {
```
Why is this part of the RpcAgent interface? It seems like this can just be a static utility function in rpc/utils.cpp?

zhaojuanmao replied:
synced up offline, this is because we want to avoid introducing python dependency for files who call the getTypeResolver(), will create two new files: one file will have virtual getTypeResolver() interface that can be called by callers without python dependency, another file will have getTypeResolver() implementation that could have python dependency

pritamdamania87 commented on torch/csrc/jit/pickle.h:
```diff
@@ -69,7 +69,8 @@ TORCH_API IValue pickle_load(const std::vector<char>& data);
 TORCH_API IValue unpickle(
```
Can we add some comments to the pickle and unpickle APIs in terms of what the type_resolve is?

pritamdamania87 commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -126,6 +129,12 @@ void Pickler::pushIValueImpl(const IValue& ivalue) {
+    #ifdef USE_DISTRIBUTED
+      pushRRef(ivalue);
+    #else
+      TORCH_INTERNAL_ASSERT(false, "RRef pickling is only supported with the distributed package");
```
Use TORCH_CHECK here or C10_THROW_ERROR. TORCH_INTERNAL_ASSERT will show up as a bug to the user and ask them to file a gh issue.

Also, what was the reason we needed to add this #ifdef? If USE_DISTRIBUTED is false, the user can't even create RRefs if I understand correctly, so how will they even pickle/unpickle one?

zhaojuanmao replied:
synced up offline, USE_DISTRIBUTED is needed here as for some use cases, users do not need to compile rref_context.h if they are not building distributed package

pritamdamania87 commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -145,6 +154,28 @@ void Pickler::pushDevice(const IValue& ivalue) {
   }
 }
 
+#ifdef USE_DISTRIBUTED
```
Why do we need this ifdef?

zhaojuanmao replied:
same as above

pritamdamania87 commented on torch/csrc/jit/pickler.h:
```diff
@@ -152,6 +152,9 @@ class Pickler {
   void pushDevice(const IValue& ivalue);
+  #ifdef USE_DISTRIBUTED
+    void pushRRef(const IValue& ivalue);
+  #endif
```
Why do we need ifdef here? This is just a function declaration.

zhaojuanmao replied:
same as above

pritamdamania87 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -1729,42 +1740,89 @@ def test_function_not_on_callee(self):
+        def return_rref(rref_var):
+            # type: (RRef[Tensor]) -> RRef[Tensor]
+            return rref_var
+
```
Nice unit tests!

pritamdamania87 reviewed (changes requested):
Requesting changes mostly for the RpcAgent interface changes.

suo commented:
Okay I am going to resign from reviewing until the other reviews are sorted out, just to avoid too much cross-chatter. Please re-request when you would me to take a look.

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

suo commented on torch/csrc/jit/unpickler.h:
```diff
@@ -19,14 +23,20 @@ class Unpickler {
+  // class_resolver is to resolve strong class type, type_resolver_ is
+  // to resolve any JIT type. class_resolver and type_resolver are not merged
+  // here because some use cases need to get strong class type that
+  // type_resolver_ can not return.
```
I don't think we should have a separate `ClassResolver` and `TypeResolver`, since they do the same job (take a type name and return the TypePtr). Instead, we should modify `StrongTypePtr` to have the `cu_` field be nullable (and enforce that if the `type_` field is a class, then `cu_` must be non-null). That way, we can merge the two resolvers.

Aside: the name of merged resolver should be `TypeResolver`, since that best describes it.

zhaojuanmao replied:
sounds good!

suo reviewed (commented):
I have one simplification suggestion. Otherwise, if @driazati signed off on the actual pickling/unpickling code, I think the JIT side is good to go.

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -159,15 +161,19 @@ class TORCH_API RpcAgent {
   const std::unique_ptr<RequestCallback> cb_;
   std::atomic<std::chrono::milliseconds> rpcTimeout_;
   std::atomic<bool> profilingEnabled_;
+  std::shared_ptr<TypeResolver> typeResolver_;
```
nit: The typeResolver_ and all the get/set methods can be static right?

pritamdamania87 reviewed (approved):
Approving since @suo mentioned the JIT side is good to go.

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "jit pickling rref"


in rpc torch script call path, we need to pickle/unpickle rref, this diff is added to make jit pickler/unpickler be able to pickle/unpickle rref. It is similar to what is implemented for PyRef::pickle() and PyRef::unpickle()

Differential Revision: [D19713293](https://our.internmc.facebook.com/intern/diff/D19713293/)

[ghstack-poisoned]
